tpm2/pcr_profile.go

```diff
@@ -87,85 +93,434 @@ type pcrProtectionProfileExtendPCRInstr struct {
 	value tpm2.Digest
 }
 
-type pcrProtectionProfileAddProfileORInstr struct {
-	profiles []*PCRProtectionProfile
+type pcrProtectionProfileBranchPointInstr struct {
+	branches []*PCRProtectionProfileBranch
 }
 
-// pcrProtectionProfileEndProfileInstr is a pseudo instruction to mark the end of a branch.
-type pcrProtectionProfileEndProfileInstr struct{}
+// pcrProtectionProfileEndBranchInstr is a pseudo instruction to mark the end of a branch.
+type pcrProtectionProfileEndBranchInstr struct{}
 
 // pcrProtectionProfileInstr is a building block of PCRProtectionProfile.
 type pcrProtectionProfileInstr interface{}
 
-// PCRProtectionProfile defines the PCR profile used to protect a key sealed with SealKeyToTPM. It contains a sequence of instructions
-// for computing combinations of PCR values that a key will be protected against. The profile is built using the methods of this type.
+type pcrProtectionProfileInstrList []pcrProtectionProfileInstr
+
+// PCRProtectionProfileBranchPoint represents a point in a parent branch
+// in which sub-branches can be inserted and populated, in order to create
+// compound policies that correspond to multiple conditions.
+type PCRProtectionProfileBranchPoint struct {
+	profile *PCRProtectionProfile
+	parent  *PCRProtectionProfileBranch
+	instr   *pcrProtectionProfileBranchPointInstr
+
+	done bool
+}
+
+func (p *PCRProtectionProfileBranchPoint) removeBranch(b *PCRProtectionProfileBranch) {
+	for i, c := range p.instr.branches {
+		if c == b {
+			if i < len(p.instr.branches)-1 {
+				copy(p.instr.branches[i:], p.instr.branches[i+1:])
+			}
+			p.instr.branches = p.instr.branches[:len(p.instr.branches)-1]
+			break
+		}
+	}
+}
+
+// AddBranch creates and returns a PCRProtectionProfileBranch corresponding to a
+// new sub-branch.
+//
+// Note that each branch created from this branch point must explicitly define
+// values for the same set of PCRs. It is not possible to generate policies where
+// each branch defines values for a different set of PCRs.
+func (p *PCRProtectionProfileBranchPoint) AddBranch() *PCRProtectionProfileBranch {
+	b := &PCRProtectionProfileBranch{profile: p.profile, parent: p}
+
+	if p.done {
+		p.profile.fail(1, "cannot add a branch to a branch point that has already been terminated")
+	}
+	if p.profile.failed() {
+		return b
+	}
+
+	p.instr.branches = append(p.instr.branches, b)
+	return b
+}
+
+func (p *PCRProtectionProfileBranchPoint) endBranchPointInternal(skip int) *PCRProtectionProfileBranch {
+	if p.done {
+		p.profile.fail(skip+1, "cannot terminate a branch point more than once")
+	}
+	if p.profile.failed() {
+		return p.parent
+	}
+
+	var branchesToEnd []*PCRProtectionProfileBranch
+	branchPoints := []*PCRProtectionProfileBranchPoint{p}
+
+	for len(branchPoints) > 0 {
```

Comment:
Supports rather than expects - because of the way that the code in efi/secureboot_policy.go works, it helps to simplify it a bit if terminating a branch point implicitly terminates the sub-branches that descend from it.